spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java

```diff
@@ -0,0 +1,206 @@
+  }
+
+  static OrcValReader<?> timestampTzs() {
+    return TimestampTzReader.INSTANCE;
+  }
+
+  static OrcValReader<?> struct(
+      List<OrcValReader<?>> readers, Types.StructType struct, Map<Integer, ?> idToConstant) {
+    return new StructReader(readers, struct, idToConstant);
+  }
+
+  static OrcValReader<?> array(OrcValReader<?> elementReader) {
+    return new ArrayReader(elementReader);
+  }
+
+  static OrcValReader<?> map(OrcValReader<?> keyReader, OrcValReader<?> valueReader) {
+    return new MapReader(keyReader, valueReader);
+  }
+
+  private static class ArrayReader implements OrcValReader<ArrayData> {
+    private final OrcValReader<?> elementReader;
+    private final List<Object> reusedList = Lists.newArrayList();
+
+    private ArrayReader(OrcValReader<?> elementReader) {
+      this.elementReader = elementReader;
+    }
+
+    @Override
+    public ArrayData nonNullRead(ColumnVector vector, int row) {
+      reusedList.clear();
+      ListColumnVector listVector = (ListColumnVector) vector;
+      int offset = (int) listVector.offsets[row];
+      int length = (int) listVector.lengths[row];
+      for (int c = 0; c < length; ++c) {
+        reusedList.add(elementReader.read(listVector.child, offset + c));
+      }
+      return new GenericArrayData(reusedList.toArray());
+    }
+  }
+
+  private static class MapReader implements OrcValReader<MapData> {
+    private final OrcValReader<?> keyReader;
+    private final OrcValReader<?> valueReader;
+
+    private final List<Object> reusedKeyList = Lists.newArrayList();
+    private final List<Object> reusedValueList = Lists.newArrayList();
+
+    private MapReader(OrcValReader<?> keyReader, OrcValReader<?> valueReader) {
+      this.keyReader = keyReader;
+      this.valueReader = valueReader;
+    }
+
+    @Override
+    public MapData nonNullRead(ColumnVector vector, int row) {
+      reusedKeyList.clear();
+      reusedValueList.clear();
+      MapColumnVector mapVector = (MapColumnVector) vector;
+      int offset = (int) mapVector.offsets[row];
+      long length = mapVector.lengths[row];
+      for (int c = 0; c < length; c++) {
+        reusedKeyList.add(keyReader.read(mapVector.keys, offset + c));
+        reusedValueList.add(valueReader.read(mapVector.values, offset + c));
+      }
+
+      return new ArrayBasedMapData(
+          new GenericArrayData(reusedKeyList.toArray()),
+          new GenericArrayData(reusedValueList.toArray()));
+    }
+  }
+
+  static class StructReader extends OrcValueReaders.StructReader<InternalRow> {
+    private final int numFields;
+    private final InternalRow internalRow;
+
+    protected StructReader(List<OrcValReader<?>> readers, Types.StructType struct, Map<Integer, ?> idToConstant) {
+      super(readers, struct, idToConstant);
+      this.numFields = readers.size();
+      internalRow = createOrReuse();
+    }
+
+    @Override
+    protected InternalRow createOrReuse() {
+      return new GenericInternalRow(numFields);
```

Comment:
My guess would be this is what is tanking the perf. Spark[Avro|Parquet]Reader seem to reuse this object. Doing the same for ORC results in test failures

Reply:
I rerun the ORC jmh tests and do see any difference.

Reply:
Do or do not?

Reply:
do not see any differece :)

Reply:
The Spark/Parquet reader doesn't currently reuse containers.

Reply:
@rdblue . Not sure I follow  this..
> The Spark/Parquet reader doesn't currently reuse containers.

I did the JMH testing for ORC before and after my patch